src/backend/distributed/executor/multi_utility.c

```diff
@@ -3787,3 +3804,78 @@ ProcessDropTableStmt(DropStmt *dropTableStatement)
 		}
 	}
 }
+
+
+/*
+ * ShouldExecuteAlterTableSequentially checks if the given ALTER TABLE
+ * statements should be executed sequentially if there is a foreign
+ * constraint from a distributed table to a reference table.
+ * In case of a column related ALTER TABLE operation, we check explicitly
+ * if there is a foreign constraint on this column from/to a reference table.
+ */
+static bool
+ShouldExecuteAlterTableSequentially(Oid relationId, AlterTableCmd *command)
+{
+	AlterTableType alterTableType = command->subtype;
+	if (alterTableType == AT_DropConstraint)
+	{
+		char *constraintName = command->name;
+		if (ConstraintIsAForeignKeyToReferenceTable(constraintName, relationId))
+		{
+			return true;
+		}
+	}
+	else if (alterTableType == AT_DropColumn || alterTableType == AT_AlterColumnType)
+	{
+		char *affectedColumnName = command->name;
+
+		/*
+		 * We need to set multi_shard_modify_mode to sequential since SET DATA TYPE
+		 * command issues one select and one DDL to see if there is incompatible
+		 * data for the new column type.
+		 */
+		if (IsTransactionBlock() && alterTableType == AT_AlterColumnType)
```

Comment:
Why do we `SetLocalMultiShardModifyModeToSequential ()` before knowing about `ForeignKeyExistsToReferenceTableOnColumn ()` ?

Also, if I execute the following where there is no reference table/fkey, I get an error

```SQL
 alter table users_table  alter COLUMN value_1 SET DATA TYPE int;
ERROR:  invalid attnum: 0
```